### PR TITLE
Fallback to .init_array when no arguments are available on glibc Linux

### PR DESCRIPTION
**Pull request by leo60228**
Linux is one of the only platforms where `std::env::args` doesn't work in a cdylib.

### CODE REVIEW

**Comment by joshtriplett** [comment]
You can't always assume that `/proc` is mounted (for instance, in a container environment or other types of embedded Linux uses). I don't think we should parse that to get arguments.

This is a known limitation of Linux, that you can't reliably get the arguments except by saving them in the main entry point. I don't think we should try to work around that. (And honestly, libraries *should* ask the application to pass in the arguments the library should use, which might not be the original arguments of the program.)

**Comment by joshtriplett** [comment]
I would personally propose closing this, unless a method arises to get the arguments out of the C library or similar.

File: src/libstd/sys/unix/args.rs [hunk 1]
```diff
@@ -46,8 +46,63 @@ impl DoubleEndedIterator for Args {
...
+    use crate::ffi::{CStr, OsString};
+    use crate::fs;
+    use crate::marker::PhantomData;
+    use super::Args;
+
+    use crate::sys_common::mutex::Mutex;
+
+    static mut ARGC: isize = 0;
+    static mut ARGV: *const *const u8 = ptr::null();
+    // We never call `ENV_LOCK.init()`, so it is UB to attempt to
```
Comment (bjorn3) [inline src/libstd/sys/unix/args.rs]: *LOCK

Comment (leo60228) [reply src/libstd/sys/unix/args.rs]: I just copied the existing code for non-macOS Unix.

**Comment by leo60228** [comment]
You can put an `extern "C" fn(
    libc::c_int,
    *const *const libc::c_char,
    *const *const libc::c_char,
)` pointer in the `.init_array` section and it'll get called with argc, argv, and envp. This should be more portable, but I'm not sure the best way to implement it in std.

**Comment by leo60228** [comment]
This is implemented in both glibc and bionic, and likely other libcs.

**Comment by leo60228** [comment]
There's also this approach that assumes `__environ` is a pointer to the stack (which is true on all major libcs), and uses the fact that arguments are always directly before the environment on the stack. I've confirmed this is the case on x86 and x86_64, and it's probably the case on other architectures. https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=732b382e6f91c3faa6e6a3a8d2a23c99

**Comment by leo60228** [comment]
@joshtriplett What do you think about these options? Using `.init_array` only works on glibc, but walking the stack is both hacky and non-constant-time.

**Comment by leo60228** [comment]
I've implemented these at https://docs.rs/libargs.

**Comment by joshtriplett** [comment]
On November 20, 2019 6:02:33 AM PST, leo60228 <notifications@github.com> wrote:
>@joshtriplett What do you think about these options? Using
>`.init_array` only works on glibc, but walking the stack  is both hacky
>and non-constant-time.

Walking the stack seems like a bad idea, considering that it's legal to change the stack before loading a library, or for that matter call the library on a thread other than the initial thread. And you'd have no way to know if you got valid data.

`.init_array` seems like a reasonable thing for the glibc target to try, given that you'll know if it got called or not. And if someone strips out the `.init_array` section, this fails gracefully. Only do this in the glibc target, though, and we need to document it so that library builders know about it.

**Comment by leo60228** [comment]
Where should this be documented?

**Comment by joshtriplett** [comment]
On November 21, 2019 5:21:30 AM PST, leo60228 <notifications@github.com> wrote:
>Where should this be documented?

In the documentation of the args functions.

**Review by joshtriplett — CHANGES_REQUESTED** [review]
Can you please add a platform-specific test for this behavior, so we know if we've done something that makes it stop working?

**Comment by leo60228** [comment]
> In the documentation of the args functions.
In `std::env` or `std::sys::unix::args`?

**Review by joshtriplett — DISMISSED** [review]
One other thought: if we initialize arguments this way, we don't *also* need to initialize them in our main function in binaries. Can you compile out that code on linux-gnu?

**Comment by joshtriplett** [comment]
Can you confirm that this works in statically linked libraries and statically linked binaries (with static glibc)?

**Comment by leo60228** [comment]
If the `sys::args::init` is disabled in `rt::init`, wouldn't that make a separate test unnecessary? If not, how can I write that test?

**Comment by leo60228** [comment]
> Can you confirm that this works in statically linked libraries and statically linked binaries (with static glibc)?

I've been having stability issues with my system, and compiling rustc causes a freeze, so I've been using `./x.py test --stage 0 --no-doc src/libstd`. Is there a way to use this libstd to compile a program?

**Comment by leo60228** [comment]
```
2019-11-21T16:28:00.3027299Z tidy error: /checkout/src/libstd/rt.rs:44: platform-specific cfg: cfg(not(all(target_os = "linux", target_env = "gnu")))
```
Why is this an error?

**Comment by leo60228** [comment]
Oh, it's because cfgs need to be in `sys`. Should I just keep executing `args::init` no matter what, then? That what it'd still work even if `.init_array` gets removed (like you mentioned).

File: src/libstd/sys/unix/args.rs [hunk 2]
```diff
@@ -80,7 +80,7 @@ mod imp {
 
     #[cfg(all(target_os = "linux", target_env = "gnu"))]
     #[used]
-    #[link_section = ".init_array"]
+    #[link_section = ".init_array.00099"]
```
[the remark this answers is not in the document]
Comment (leo60228) [reply src/libstd/sys/unix/args.rs]: I was using the `ctor` crate in my library, which puts constructors in the `.ctor` section. The arguments hadn't been initialized when my code got called.

**Comment by joshtriplett** [comment]
@leo60228 
> In `std::env` or `std::sys::unix::args`?

In `std::env::args` and `std::env::args_os` where people will actually see it in the documentation.

> Oh, it's because cfgs need to be in sys. Should I just keep executing args::init no matter what, then?

How about marking `args::init` as `#[inline(always)]`, and making it a no-op on linux-gnu? Your `.init_array` code can call the "real" implementation. That way there's no `cfg` in the entry point code, and there's no duplicate initialization of arguments.

Also, investigating further, this *does* work in a binary, including a statically linked binary. Also worth noting, a glibc-based binary or library has content in `.init_array` by default, and will segfault in `__libc_csu_init` if you strip out `.init_array`, so please ignore my previous comments about the possibility of stripping out `.init_array`. Given that glibc uses this for its own purposes, using it in Rust doesn't seem unreasonable to me. It'll get left out when building for `no_std`.

So, at this point, I think we just need a test, and a bit of cfg logic to make `rt::init`'s arg initialization a no-op.

File: src/libstd/sys/unix/args.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (joshtriplett) [reply src/libstd/sys/unix/args.rs]: Ah. I don't want us to get into priority wars over what runs first before main, and it seems questionable to me to use args from ctors, but alright

**Comment by joshtriplett** [comment]
Thank you for all the work you've put into this!

@rust-lang/libs, does this seem reasonable to you? I think this seems worth doing.

**Comment by sfackler** [comment]
Seems plausible to me. I assume the `00099` is a priority order?

**Comment by leo60228** [comment]
Yes.

**Comment by Amanieu** [comment]
Note that calling `init_array` functions with `argc` and `argv` is a glibc extension. In particular, musl doesn't do this and you will have garbage in the arguments.

**Comment by Amanieu** [comment]
For musl, perhaps the best thing to do would be to parse /proc/self/cmdline. Sure, it's not 100% reliable since some programs deliberately overwrite their cmdline in memory to appear with a special name in ps/top, but at that point the arguments are no longer present in memory anyways.

**Comment by leo60228** [comment]
@Amanieu I use `target_env = "gnu"`, so the previous behavior is preserved on non-glibc.

**Comment by alexcrichton** [comment]
I would agree this seems reasonable to land, supporting argc/argv access in a cdylib is best effort on a per-platform basis, so adding platform specific code to support it somewhere seems fine by me.

**Comment by joshtriplett** [comment]
What's the libs process here? Should this go through rfcbot for libs, or since it doesn't add new API, is the current expressed support sufficient?

@leo60228 Could you please rebase -i into a set of logical commits (that don't include reverts or fixes to previous patches, for instance)?

**Comment by JohnCSimon** [comment]
Ping from triage 
@leo60228 Could you please rebase this PR as per the change request? 
Thank you.

**Review by dtolnay — APPROVED** [review]
This looks good to me, thanks @leo60228!

**Comment by sdroege** [comment]
FYI this is actually not safe like this, see https://github.com/rust-lang/rust/issues/105999

**Pull request merged**